Use a dedicated thread and event queue to invoke the event handlers in AriClient.

pragmatrix opened the pull request:
This pull request should fix #11 and is compatible with the current AriClient interface.

Some notes and concerns:

- I've added `IDisposable` interface and a `Dispose()` method to AriClient that invokes `Disconnect()` so that it is obvious from the interface, that some cleanup needs to be done. 
- The thread / event queue is shut down in `Disconnect()`, but will be kept when an automatic reconnect happens.
- The thread will keep running, if neither `Disconnect()` nor `Dispose()` is called. I found no reliable way to cancel the thread so far. The finalizer won't be called, because the thread must refer to the AriClient instance to deliver the events. One solution for that would be to deliver the AriClient instance with each event.
- I've also put a lock around all members that may be changed from different threads.
- I am a bit concerned that the event handler invocations are serialized now. Client code that blocks inside an event handler may behave differently, or just stall.
- I did a few tests within my setup, but have not tested the examples.

<!-- Reviewable:start -->
[<img src="https://reviewable.io/review_button.png" height=40 alt="Review on Reviewable"/>](https://reviewable.io/reviews/skrusty/asternet.ari/12)
<!-- Reviewable:end -->

skrusty commented:
I think the new Dispatcher should be interfaced to allow previous behaviour to be implemented.
Although it would have been daft, in theory someone switching to a newer version of AsterNET.ARI with the new dispatched could easily deadlock their code by waiting for an event from within an event.

This ins't like AsterNET's AMI implementation mind you where the deadlock was down to the connecitonmanager being single threaded, but in this case the dispatcher would not serve up the waited for event.

pragmatrix commented:
I've introduced an `EventDispatchingStrategy` property to AriClient, then I added a thread pool dispatcher, and made the dedicated thread dispatcher the default.

I've decided against making the interface to the dispatchers public for now, mainly because the dispatcher may be instantiated multiple times in AriClient. If we want the dispatcher completely pluggable, we could make the `CreateDispatcher()` method virtual.

skrusty commented:
Looks good, can we make the default the threadpool please, that way we're 100% backward compatible and upgrading wont require code change.

Make sense? Thanks.

pragmatrix commented:
Thread pool is now the default and also exceptions in the event handlers are getting reported and then ignored, like it was before.

jonasswiatek commented:
I think it looks good. I've mainly reviewed the new producer/consumer code and I can't immediately see any flaws with it. Even implements cancellation and everything! Top notch @pragmatrix !

pragmatrix commented:
Thank you for the review @jonasswiatek!
